starknet-client/src/helpers.rs

```diff
@@ -29,11 +24,97 @@ const UPDATE_WALLET_EXTERNAL_TRANSFER_LEN: usize = 4;
 
 /// Error message emitted when a public blinder share is not found in calldata
 const ERR_BLINDER_NOT_FOUND: &str = "public blinder share not found in calldata";
-/// Error message emitted when a blob encoding is invalidly structured
-const ERR_INVALID_BLOB_ENCODING: &str = "blob encoding incorrect";
 /// Error message emitted when an invalid selector is given in the transaction's execution trace
 const ERR_INVALID_SELECTOR: &str = "invalid selector received";
 
+// --------------------
+// | Calldata Parsing |
+// --------------------
+
+// InvokeV1 calldata is structured as follows:
+//  0: `call_array_len`
+//  1..4:
+//      1: `contract_addr`
+//      2: `selector`
+//      3. `data_offset`
+//      4. `data_len`
+//  ... for `call_array_len` elements
+//  `calldata_array_len` * 4 + 1: `calldata_len`
+//  ... `raw_calldata`
+/// The index of the `call_array_len` argument in InvokeV1 calldata
+const CALL_ARRAY_LEN_IDX: usize = 0;
+/// The length of each call array element's metadata in InvokeV1 calldata
+const CALL_ARRAY_ELEMENT_METADATA_LEN: usize = 4;
+/// The index of the contract address argument in a call array element's metadata
+const CALL_ARRAY_CONTRACT_ADDR_IDX: usize = 0;
+/// The index of the selector argument in a call array element's metadata
+const CALL_ARRAY_SELECTOR_IDX: usize = 1;
+/// The index of the data offset argument in a call array element's metadata
+const CALL_ARRAY_DATA_OFFSET_IDX: usize = 2;
+/// The index of the data length argument in a call array element's metadata
+const CALL_ARRAY_DATA_LEN_IDX: usize = 3;
+
+/// Parses the first darkpool transaction in a call array
+///
+/// n.b. It is generally assumed that only one darkpool transaction exists in a given
+/// call array, and the caller should take care to ensure this is the case
+///
+/// Returns the selector and calldata of the first darkpool transaction found
+pub(crate) fn parse_first_darkpool_transaction(
+    tx_calldata: Vec<StarknetFieldElement>,
+    darkpool_contract_addr: &StarknetFieldElement,
+) -> Option<(StarknetFieldElement, Vec<StarknetFieldElement>)> {
+    // Parse the number of calls in the call array
+    let n_calls = starknet_felt_to_usize(&tx_calldata[CALL_ARRAY_LEN_IDX]);
+
+    // Find the first renegade transaction in the call array metadata
+    let metadata_start = CALL_ARRAY_LEN_IDX + 1;
+    let metadata_end = metadata_start + (n_calls * CALL_ARRAY_ELEMENT_METADATA_LEN);
+
+    let (selector, data_offset, data_len) = find_first_darkpool_transaction(
+        &tx_calldata[metadata_start..metadata_end],
+        darkpool_contract_addr,
+    )?;
+
+    // Parse the calldata of the transaction
+    let calldata_start = metadata_end;
+    let calldata = tx_calldata[calldata_start..].to_vec();
+    let tx_calldata = calldata[data_offset..(data_offset + data_len)].to_vec();
```

Comment:
nit: bouncing between `let calldata = tx_calldata[...]` and `let tx_calldata = calldata[...]` is a bit tough to read, could we rename the second var 😅